### DISCUSSION TITLE
list of parts

### DISCUSSION THREAD

**Discussion by Gene0Grey** [ZeroZeroOne-dev/00Key#6] [Q&A] [answered]
Hi I just wanted to ask, you have said the mosfet is IRL540N with a 10k ohm resistor I assume the the diodes are 1N4148 and the resistors are 560 ohms. The firmware how is that flashable to microcontroller as there is no documentation for flashing also idk much about flashing a teensy as i have never worked with them before thank you.

**Comment by ZeroZeroOne-dev** [comment, marked as answer]
All the components you said are correct, but to be sure here is a list: 

- 1 SSD1306 OLED screen; I used [this one](https://www.aliexpress.com/item/32638662748.html?spm=a2g0s.9042311.0.0.14ca4c4dpIS7N0) from aliexpress. If you use a different one, make sure its I2C and the pins are in the same order.
- 84 diodes: the most common one, 1N4148, are perfect and they are readily available.
- 84 LEDs: You can use whatever color you want, they just have to fit in the keyboard switches (so get some without a flange). I used [these ones](https://www.tme.eu/be/en/details/osw5dk3e31b/tht-leds-3mm/optosupply/).
- 84 resistors: The value of your resistors depends on what LED you picked, you can use an online tool to figure that out. You only want 3 or 4 mA for each LED or its gonna be too bright. For the LEDs I used the resistor value is 560Ω.
- 1 n-channel logic level mosfet: The IRL540N are pretty common.
- 1 gate resistor for the mosfet: This is just a pulldown resistor, any above 1KΩ should be fine I use a 10KΩ one.
- 1 Teensy 4.1
- 84 of your keyboard switches of choice as long as they allow for LEDs (MX or Alps style, although I personally have not tested Alps).
- 1 set of stabilizers: 3 x 2U and 1x 6.25U.

For compiling and modifying the firmware, i recommend using Visual Studio code with the [PlatformIO extension](https://marketplace.visualstudio.com/items?itemName=platformio.platformio-ide) (that should pick up the project in the firmware folder automatically). That will give you easy actions for building and deploying.

**Reply by Gene0Grey** [reply]
hello sir thank you for confirming the parts I have already ordered the pcb also I have one more question is there like a guide or documentation to flash the firmware as I have no idea what to do with the firmware files.so it would be nice to get like guide or a video to do it,
 thank you

**Reply by ZeroZeroOne-dev** [reply]
Building and uploading the file to the device is pretty easy, once you have set up PlatformIO its just one button. 
(Its as easy as just opening the firmware folder, PlatformIO will recognize it as a project and offer you the build and upload buttons).

There are a lot of guides about PlatformIO out there, but [this ](https://youtu.be/sm6QxJkWcSc)is a nice short one to get started. 

However if you want it I can compile the firmware for you and upload the hex, you just have to use the [Teensy Loader Application](https://www.pjrc.com/teensy/loader.html) to upload the hex to the device.

**Reply by Gene0Grey** [reply]
oh yes that would be very nice and easy for everyone actually having the hex file is one click upload so it would be nice

**Reply by ZeroZeroOne-dev** [reply]
Little update: I'm planning to set up a github action to automatically compile the hex file. 
If I haven't done that by the time you need the hex file, just send me a message here and I'll make sure you have it.

**Reply by mjparme** [reply]
FWIW, after plugging the teensy into my computer I just went into the firmware directory and executed `pio run --target upload` and then platformIO did its thing. It downloaded a few libraries, compiled the code, and then it uploaded the firmware to the teensy.

That was faster than firing up VSCode or CLion.
